client/in_memory_provider.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2022 Lekko Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/lekkodev/go-sdk/internal/memory"
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/known/structpb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+type InMemoryProviderOptions struct {
+	// API key used to communicate with Lekko.
+	// lekko_**********
+	APIKey string
+	// Repository Key of the configuration repository you wish to read
+	// configs from.
+	RepositoryKey RepositoryKey
+	// Interval at which you want to refresh configs from Lekko.
+	// Recommend 15 to 30 seconds.
+	UpdateInterval time.Duration
+}
+
+type inMemoryProviderType string
+
+const (
+	inMemoryProviderTypeBackend inMemoryProviderType = "backend"
+
+	minUpdateInterval = time.Second
+)
+
+// Constructs a provider that refreshes configs from Lekko backend repeatedly in the background,
+// caching the configs in-memory.
+func BackendInMemoryProvider(ctx context.Context, opts *InMemoryProviderOptions) (Provider, error) {
```

Comment:
should this be after definition of `inMemoryProvider` (line 83+)?

Reply:
order of definition doesn't matter, and i'd rather keep the public, client-facing methods to the top and the private ones below

Reply:
I just realized that `inMemoryProvider` itself is private :) Which is something very go-specific I guess. I was just confused to have "constructor" defined before the actual struct :)